# **Textile Company Data Analysis**

## **Summary**

This notebook walks through a full data analysis for a textile company’s business deals. We focus on predicting commissions and finding ways to improve business results. The analysis covers more than 4,500 contracts to spot what drives success, strengthen supplier and customer relationships, and build models that predict commissions with 98.4% accuracy.

**Main Questions We Answer:**
- Which suppliers and customers help the business the most?
- How can we predict commission amounts for new deals?
- What factors affect profits the most?

**About the Data:**
- **Company**: Textile Company (manufacturing and exporting textiles)
- **What’s Included**: Deals with suppliers, customers, and internal teams
- **Time Frame**: Several years, covering different market conditions
- **Size**: Over 4,500 contracts
- **Key Details**: Commission amounts, payment terms, exchange rates, quantities, and performance

## **Business Background & Goals**

**Who’s Involved:**
- **Suppliers**: Provide materials and manufacturing (e.g., SWU, Samad Rubber)
- **Customers**: International buyers (e.g., Tiffosi, Takko)
- **Internal Teams**: Managed by experienced leaders (names kept private)

**What We Want to Achieve:**
1. **Find What Works**: Discover which relationships and patterns lead to success
2. **Predict Commissions**: Build models to forecast commissions for planning
3. **Clean Data**: Make sure our data is accurate and reliable
4. **Give Useful Advice**: Share clear recommendations for growth and efficiency

## **What’s in This Notebook**

### **Step 1: Data Preparation**
- Clean and organize the data
- Spot and fix errors or outliers
- Rename columns for clarity

### **Step 2: Business Insights**
- Explore the data to find important trends
- Map out relationships between suppliers, customers, and teams
- Analyze commission structures and payment terms

### **Step 3: Predictive Modeling**
- Build machine learning models to predict commissions
- Achieve high accuracy (98.4%)
- Find out which features matter most

### **Step 4: Recommendations**
- Share practical tips for improving operations
- Suggest how to allocate resources based on data
- Point out risks and opportunities

---

*This analysis turns raw business data into clear insights, helping the company make smart decisions and grow sustainably.*


### **Environment Setup & Library Import**

#### **What We're Doing:**
We're getting our workspace ready with the main Python tools for working with data, making charts, and building machine learning models.

**Main Libraries:**
- **pandas**: For handling and analyzing data
- **seaborn & matplotlib**: For making charts and graphs
- **scikit-learn**: For machine learning and data prep
- **plotly**: For interactive and business-friendly visuals

**Extra Tools:**
- **IPython.display**: For better notebook formatting
- **warnings**: To keep our output clean and easy to read

In [1]:
import pandas as pd
import numpy as np

# Enhanced Display and Widgets
from IPython.display import display, HTML, Markdown
import ipywidgets as widgets
from IPython.display import clear_output


print(" All libraries imported successfully!")
print(" Environment ready for comprehensive analysis!")

 All libraries imported successfully!
 Environment ready for comprehensive analysis!


### *Load the dataset*

```python

In [3]:
df = pd.read_csv('textile_sales.csv')

### **Column Renaming: Making Things Clear**

#### Why We Rename Columns

Renaming columns helps make the data easier to understand and work with. Here’s what we changed:

- **Expected_Payment → Expected_Payment_Date**: Shows this is a date
- **Rate → Per_item_Rate(PKR/$)**: Tells us the currency and unit
- **Dollar_Rate → Per_item_Rate($)**: Makes it clear this is in USD
- **Bal → Balance**: Says exactly what it means
- **Exchange → Exchange(PKR_to_$)**: Explains the conversion
- **Received → Commission_after_Tax(PKR)**: Shows it’s the net commission

These updates make the data more readable and connect the columns to real business terms, so analysis is smoother and more meaningful.

In [4]:
df = df.rename(columns={
    'Expected_Payment': 'Expected_Payment_Date ',
    'Rate': 'Per_item_Rate(PKR/$)',
    'Dollar_Rate': 'Per_item_Rate($)',
    'Bal': 'Balance',
    'Exchange': 'Exchange(PKR_to_$)',
    'Received': 'Commission_after_Tax(PKR) ',
})

print("Column renaming completed!")
print("Dataset now has more intuitive and business-friendly column names")

Column renaming completed!
Dataset now has more intuitive and business-friendly column names


<h2 style="
    font-family: 'Orbitron', monospace;
    font-size: 3rem;
    font-weight: 900;
    background: linear-gradient(45deg, #00ffff, #0080ff, #8000ff, #ff0080);
    background-size: 400% 400%;
    -webkit-background-clip: text;
    -webkit-text-fill-color: transparent;
    background-clip: text;
    animation: gradientShift 4s ease-in-out infinite;
    margin-bottom: 30px;
    text-shadow: 0 0 30px rgba(0, 255, 255, 0.5), 0 8px 32px #8000ff;
    letter-spacing: 2px;
    transition: all 0.6s cubic-bezier(0.4, 0, 0.2, 1);
    display: inline-block;
">
EDA (Exploratory Data Analysis)
</h2>

<style>
@keyframes gradientShift {
    0%, 100% { background-position: 0% 50%; }
    50% { background-position: 100% 50%; }
}
</style>

### **1. Data Cleaning – Laying the Groundwork**

#### **What We're Doing:**
We're starting by cleaning up the data to make sure it's accurate and ready for analysis. Good data is the key to getting useful results.

### **Why Data Cleaning Matters:**
- **Accuracy**: Clean data leads to better insights and predictions
- **Consistency**: Standard formats make analysis easier
- **Reliability**: Removing mistakes and odd values keeps results trustworthy
- **Efficiency**: Well-organized data speeds up our work

### **Managing Columns**
#### **Removing Redundant Information**

**Issue Found:** Some columns mean the same thing, so we can drop a few to keep things simple:

**Redundant Columns:**
- **`Per_item_Rate(PKR/$)` vs `Per_item_Rate($)`**: Both show item prices, just in different currencies. Since we have exchange rates, one is enough.
- **`SW_After_Tax_Value` vs `SW_Rcvd_after_Tax`**: Both show the commission received after tax.

**Why Remove Them:**
- **Avoid Confusion**: Duplicate columns can confuse analysis and models
- **Keep It Clear**: Fewer columns make results easier to understand
- **Work Faster**: Less data means quicker processing

> **Solution:** Keep the columns that are clear and useful, and remove the duplicates.


In [4]:
df = df.drop(columns=[
    'Per_item_Rate(PKR/$)',
    'SW_After_Tax_Value',
])

print(" Redundant features removed successfully!")
print(f" Dataset size reduced from {df.shape[1] + 2} to {df.shape[1]} columns")
print(" Improved data quality and model readiness")

 Redundant features removed successfully!
 Dataset size reduced from 26 to 24 columns
 Improved data quality and model readiness


#### **Check Types in the Categorical Features**

In [5]:
columns = ['Supplier','Customer','Style_Name']

for col in columns:
    display(HTML(f"<h4 style='color: #2a7ae2;'>Value counts for <b>{col}</b>:</h4>"))
    vc = df[col].value_counts().reset_index()
    vc.columns = [col, 'Count']
    display(vc)

Supplier  Count
0              SWU   1443
1     Samad Rubber    363
2             ASCO    242
3      AR Apparels    208
4    MR Industries    150
..             ...    ...
99        Sportler      1
100             CW      1
101          Union      1
102          Delta      1
103       Artistic      1

[104 rows x 2 columns]

Customer  Count
0                 Tiffosi   1257
1                   Takko    825
2              Takko Knit    487
3               Takko Fab    439
4   Authentic Style knits    251
..                    ...    ...
76           Jomo Fashion      1
77            Coccodrillo      1
78           Big Star Fab      1
79                 Bizbee      1
80               Big Star      1

[81 rows x 2 columns]

Style_Name  Count
0                                                 TKMEN    321
1                                                 TKLAD    299
2                                                 TKKID    152
3                                                 TKMND     40
4     E-5124F-Blue, 10.5 Oz, 3/1 RHT, Rain Stretch, ...     37
...                                                 ...    ...
2463                                      E-5124 FOD/OB      1
2464                                             Zoe_25      1
2465                                             Zoe_24      1
2466                                          John_K363      1
2467                          D10608BC02802A1NO(Repeat)      1

[2468 rows x 2 columns]

### **Spotting Outliers and Checking for Errors**

#### **What We're Doing:**
We're looking for numbers in the data that don't make sense or break business rules, and then removing them.

**How We Do It:**
1. **Check the Stats**: Find values that look odd or stand out
2. **Use Business Know-How**: Make sure numbers fit real-world expectations
3. **See the Impact**: Understand how strange values could mess up our results
4. **Remove Only the Mistakes**: Delete clear errors, but keep unusual cases that might be real

**Why This Is Important:**
- **Better Predictions**: Outliers can throw off our models
- **Smarter Decisions**: Bad data leads to bad business choices
- **Trustworthy Results**: Clean data means our reports and metrics are reliable

___

**Drop rows where Per_item_Rate($) is less than 0**

In [6]:
df = df[df['Per_item_Rate($)'] >= 0]
print("Negative item rates removed successfully!")

Negative item rates removed successfully!


**Drop rows where Contract_Qty. is less than 0**

In [7]:
df = df[df['Contract_Qty.'] >= 0]
print("Negative contract quantities also removed successfully!")

Negative contract quantities also removed successfully!


### **Checking Exchange Rates: Making Sure the Data Makes Sense**

#### **Why We Look at Exchange Rates:**

**Business Reality:**
- **Normal Range**: The US dollar usually trades between 100 and 300 PKR
- **History Check**: Even in tough times, rates don’t go way outside this range
- **Data Quality**: If we see numbers far above or below this, it’s probably a mistake

**How We Keep Data Clean:**
1. **Spot Weird Numbers**: Find exchange rates that don’t fit the usual pattern
2. **Think It Through**: Decide if it’s a real case or just a typo
3. **Fix the Problem**: Remove obvious errors, and double-check anything unusual

### **Why This Matters:**
- Helps us trust our analysis and reports
- Makes sure our business decisions are based on real numbers
- Keeps our models and predictions accurate

In [8]:
# Check the rows where the Dollar_Exchange_Rate is greater than 1
Exchange_Rate =df[df['Dollar_Exchange_Rate'] > 1]

### **Exchange Rate Outlier: Value = 155**

**What We Found:**
- There’s just one row where `Dollar_Exchange_Rate` is 155.
- This is a clear mistake—exchange rates should be small numbers like 0.0033 (for 300 PKR/USD).

**Why It’s Wrong:**
1. **Doesn’t Make Sense**: It would mean 1 PKR equals 155 USD, which is impossible.
2. **History Check**: The PKR has never been stronger than the USD.
3. **Data Consistency**: All other exchange rates look normal.

**Why Remove It:**
- It’s only one row, so we lose almost nothing.
- Removing it keeps our calculations and reports accurate.
- It stops our models from learning the wrong thing.

**Action:** We’ll delete this row to keep our data clean and reliable.


In [9]:
df = df[df['Dollar_Exchange_Rate'] <= 1]

**Dollar-Value Analysis** : Based on historical data, it has never occurred that the value of the US dollar dropped below 100 PKR during the time period covered by this dataset.

Drop the enteties that are below 100

In [10]:
df = df[df['Exchange(PKR_to_$)'] >= 100]
print("Exchange rates below 100 PKR per USD removed successfully!")

Exchange rates below 100 PKR per USD removed successfully!


**Commission Analysis** : Getting the negative commission values

In [11]:
negative_commission_df = df[(df['Commission_Dollar_Value'] < 0) |
                            (df['Commission_Percentage'] < 0) |
                            (df['Commission_PKR_Value'] < 0) |
                            (df['Commission_SW_Percentage'] < 0) |
                            (df['SW_Rcvd_after_Tax'] < 0)]

### **Commission Value Errors: What’s Going On?**

**What We Found:**  
Some rows have negative commission values in different columns.

**Examples:**  
- One row shows `Commission_Dollar_Value` as -208 (should be +208)  
- Another has `Commission_PKR_Value` as -34,055 (should be +34,055)  
- This looks like a pattern—sign errors during data entry

**Why This Is a Problem:**  
- **Commissions are earnings**—they should always be positive  
- **Negative commissions** would mean the company is paying out, which doesn’t fit the business  
- Other related fields for the same deals show positive values

**Possible Reasons:**  
- Manual entry mistakes  
- Issues when importing data  
- Calculation errors in the original system

**How It Hurts Our Analysis:**  
- Skews averages and totals  
- Confuses machine learning models  
- Gives a false picture of profits

**What We’ll Do:**  
We’ll remove these rows to keep our data accurate. Since we don’t have enough info to fix the numbers, deleting them is the safest option.

In [12]:
df = df[(df['Commission_Dollar_Value'] >= 0) &
        (df['Commission_Percentage'] >= 0) &
        (df['Commission_PKR_Value'] >= 0) &
        (df['Commission_SW_Percentage'] >= 0) &
        (df['SW_Rcvd_after_Tax'] >= 0)]

# Display the cleaned DataFrame
print("Data cleaning completed successfully!")


Data cleaning completed successfully!


### **Duplicate Row Analysis: Why It Matters**

#### **How Duplicates Happen in Business Data**

**Common Reasons for Duplicate Records:**
1. **System Sync Issues**: Data gets copied from different systems
2. **Manual Entry Mistakes**: Someone enters the same info twice
3. **Reporting Overlap**: The same deal shows up in more than one report
4. **Data Transfers**: Errors when moving data between systems
5. **API Problems**: Multiple updates create extra records

**Why Duplicates Are a Problem:**
- **Numbers Look Bigger Than They Are**: Sales and commissions get counted twice
- **Analysis Gets Confusing**: We might make wrong decisions based on bad data
- **Machine Learning Struggles**: Models learn from repeated info, which isn’t helpful
- **Wasted Resources**: Time and money go to the wrong places

**How We Find Duplicates:**
- **Exact Match**: Look for rows where every column is the same
- **Business Check**: Make sure duplicates aren’t real deals
- **Time Review**: See if duplicates happen at the same time

**What We Do:**  
We drop duplicate rows to keep our data clean and our results trustworthy.

In [13]:
# Check duplicate rows in the DataFrame
duplicate_rows = df[df.duplicated()]
if not duplicate_rows.empty:
    print("Duplicate rows found:")
    display(duplicate_rows)
else:
    print("No duplicate rows found.")

Duplicate rows found:


Team_Head    Supplier     Customer        Contract_Date  \
2182    Muneeb  Siddiqsons    Takko Fab  2022-01-14 00:00:00   
3889       SWU   Pak Denim  Tiffosi Fab  2024-02-12 00:00:00   

            Dispatch_Date Payment_Days Expected_Payment_Date  Style_Name  \
2182  2023-10-31 00:00:00           60    2023-12-30 00:00:00    DS-9359   
3889  2024-07-03 00:00:00           60    2024-09-01 00:00:00    PD-3023   

      Dollar_Exchange_Rate  Per_item_Rate($)  ... Sales_Volume  \
2182                0.0063              3.44  ...        74041   
3889                0.0040              2.46  ...        16236   

      Exchange(PKR_to_$)  Commission_Dollar_Value  Commission_Percentage  \
2182               160.0                  1480.81                   0.02   
3889               250.0                   324.72                   0.02   

     Commission_PKR_Value  Commission_SW_Percentage  SW_Commission_PKR  Tax_%  \
2182             236930.0                     0.008              94772   0.12   
3889              81180.0                     0.008              32472   0.12   

      Commission_after_Tax(PKR)   SW_Rcvd_after_Tax  
2182                      208498            83399.2  
3889                       71438            28575.0  

[2 rows x 24 columns]

In [14]:
df = df.drop_duplicates() 

In [15]:
print("Duplicate rows finally removed successfully!")
print(f"Final dataframe shape: {df.shape}")
print("Data information:")
df.info()
print("\nData statistics:")
display(df.describe())

Duplicate rows finally removed successfully!
Final dataframe shape: (4614, 24)
Data information:
<class 'pandas.core.frame.DataFrame'>
Index: 4614 entries, 0 to 4724
Data columns (total 24 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   Team_Head                   4614 non-null   object 
 1   Supplier                    4614 non-null   object 
 2   Customer                    4614 non-null   object 
 3   Contract_Date               4614 non-null   object 
 4   Dispatch_Date               4614 non-null   object 
 5   Payment_Days                4614 non-null   object 
 6   Expected_Payment_Date       4614 non-null   object 
 7   Style_Name                  4614 non-null   object 
 8   Dollar_Exchange_Rate        4614 non-null   float64
 9   Per_item_Rate($)            4614 non-null   float64
 10  Contract_#                  4614 non-null   object 
 11  Contract_Qty.               4614 non-null   float64
 12

Dollar_Exchange_Rate  Per_item_Rate($)  Contract_Qty.  \
count           4614.000000       4614.000000    4614.000000   
mean               0.843201          5.876776    9324.590811   
std                0.361364          7.877695   13526.553551   
min                0.003700          0.630000       0.000000   
25%                1.000000          3.630000    1302.750000   
50%                1.000000          5.700000    3215.500000   
75%                1.000000          7.400000   13036.250000   
max                1.000000        250.000000  260854.000000   

       Shipped_Quantity       Balance  Exchange(PKR_to_$)  \
count       4614.000000   4614.000000         4614.000000   
mean        9298.394761     35.108871          216.504593   
std        13524.800639    678.853279           55.980086   
min            0.000000  -1917.000000          100.000000   
25%         1284.000000      0.000000          160.000000   
50%         3196.500000      0.000000          206.000000   
75%        12959.250000      0.000000          278.400000   
max       260854.000000  36462.000000          416.300000   

       Commission_Dollar_Value  Commission_Percentage  Commission_PKR_Value  \
count              4614.000000            4614.000000          4.614000e+03   
mean               1715.069055               0.043508          3.823316e+05   
std                2914.819070               0.066869          6.874538e+05   
min                   0.000000               0.000000          0.000000e+00   
25%                 247.675000               0.020000          4.949600e+04   
50%                 603.480000               0.030200          1.288695e+05   
75%                1743.902500               0.050000          3.724210e+05   
max               38642.800000               1.000000          1.102672e+07   

       Commission_SW_Percentage  SW_Rcvd_after_Tax  
count               4614.000000       4.614000e+03  
mean                   0.035507       2.539001e+05  
std                    0.066573       4.524490e+05  
min                    0.000000       0.000000e+00  
25%                    0.020000       3.729925e+04  
50%                    0.025900       1.020349e+05  
75%                    0.043100       2.482658e+05  
max                    1.000000       7.277637e+06

> Now we clean the dataset by removing the identified duplicates and outliers.
on the next files, we will continue with the data analysis and modeling steps.

+ For Data Visualization (DV): 
+ For Machine Learning (ML) modeling: 

___